# HOTEL BOOKING DEMAND ANALYTICS PROJECT

## Project Objective

The goal of this project is to clean, analyze, and understand hotel booking data.

In this project, I performed data cleaning, handled missing values, investigated duplicates, fixed data quality issues, and prepared the dataset for further analysis and dashboard creation.

In [3]:
import pandas as pd

hbdc = pd.read_csv("C:\\Users\\varsh\\Desktop\\DA\\Hotel_booking_DC\\Data\\hotel_bookings_raw.csv.csv")

hbdc.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,7/1/2015
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,7/1/2015
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,7/2/2015
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,7/2/2015
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,7/3/2015


In [4]:
# To get the shape of the dataset: Row total & Column total, bcoz I need to know, how large the datatset is
hbdc.shape

(119390, 32)

In [5]:
# To get the entire info of the dataset, provides: Column names, Datatypes, Non-null counts, Memory usage. Here, I need to note which cols have missing vals & incorrect datatypes 
hbdc.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal                       

## Dataset Overview

The dataset contains booking information for both Resort Hotels and City Hotels.

Some of the information in the data includes:

- Booking status
- Lead time
- Arrival dates
- Guest information
- Room types
- Country information
- Booking channels
- Hotel pricing (ADR)
- Customer types

The dataset initially contained 119,390 records and 32 columns.

## Initial Dataset Inspection

Before starting the cleaning process, I explored the dataset to understand its structure and identify possible data quality issues.

### Observations

- Dataset contains 119,390 rows and 32 columns.
- Data includes bookings from both Resort Hotels and City Hotels.
- Some columns contain missing values.
- Reservation status date is stored as text and may need datatype conversion.

In [6]:
# Shows excatly how many missing values exist
hbdc.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

## Missing Values Investigation

After checking the dataset, I found missing values in four columns:

* Children - 4 missing values
* Country - 488 missing values
* Agent - 16,340 missing values
* Company - 112,593 missing values

### Observation

The missing values in the children and country columns are very few and can be handled.

As the agent and company columns contain a large number of missing values, I decided to investigate them further before choosing to decide anything.


In [7]:
# Counts duplicate rows
hbdc.duplicated().sum()

np.int64(31994)

## Duplicate Records Investigation

I found 31,994 duplicate records in the dataset.

Before deleting them, I checked whether they were genuine records or actually duplicates. After investigation, I confirmed that the duplicated rows contained the same information across all columns.

In [8]:
# Provides Statistics for numeric cols: Count, Mean, Std, Min, 25%, 50%, 75%, Max
hbdc.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119386.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,103050.000000,6797.000000,119390.000000,119390.000000,119390.000000,119390.000000
mean,0.370416,104.011416,2016.156554,27.165173,15.798241,0.927599,2.500302,1.856403,0.103890,0.007949,0.031912,0.087118,0.137097,0.221124,86.693382,189.266735,2.321149,101.831122,0.062518,0.571363
std,0.482918,106.863097,0.707476,13.605138,8.780829,0.998613,1.908286,0.579261,0.398561,0.097436,0.175767,0.844336,1.497437,0.652306,110.774548,131.655015,17.594721,50.535790,0.245291,0.792798
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,6.000000,0.000000,-6.380000,0.000000,0.000000
25%,0.000000,18.000000,2016.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.290000,0.000000,0.000000
50%,0.000000,69.000000,2016.000000,28.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,179.000000,0.000000,94.575000,0.000000,0.000000
75%,1.000000,160.000000,2017.000000,38.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,270.000000,0.000000,126.000000,0.000000,1.000000
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,1.000000,26.000000,72.000000,21.000000,535.000000,543.000000,391.000000,5400.000000,8.000000,5.000000


### FINDING THE OUTLIERS FROM THE DATA

In [9]:
hbdc[ hbdc["adr"] < 0 ]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
14969,Resort Hotel,0,195,2017,March,10,5,4,6,2,...,No Deposit,273.0,NaN,0,Transient-Party,-6.38,0,0,Check-Out,3/15/2017


## Outlier - 1

While exploring the dataset, I noticed one booking with a negative ADR value (-6.38).

Since hotel room rates cannot normally be negative, I marked this record for further investigation during the cleaning process.

In [10]:
hbdc[hbdc["adr"] > 1000 ]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
48515,City Hotel,1,35,2016,March,13,25,0,1,2,...,Non Refund,12.0,NaN,0,Transient,5400.0,0,0,Canceled,2/19/2016


## Outlier - 2

I noticed another booking with a very high ADR value (5400).

This value was extremely different from the rest of the dataset and appeared unrealistic.

In [11]:
hbdc[hbdc["adults"] > 10 ]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
1539,Resort Hotel,1,304,2015,September,36,3,0,3,40,...,No Deposit,NaN,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
1587,Resort Hotel,1,333,2015,September,36,5,2,5,26,...,No Deposit,96.0,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
1643,Resort Hotel,1,336,2015,September,37,7,1,2,50,...,No Deposit,NaN,NaN,0,Group,0.0,0,0,Canceled,1/18/2015
1752,Resort Hotel,1,340,2015,September,37,12,2,5,26,...,No Deposit,96.0,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
1884,Resort Hotel,1,347,2015,September,38,19,2,5,26,...,No Deposit,96.0,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
1917,Resort Hotel,1,349,2015,September,39,21,1,3,27,...,No Deposit,NaN,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
1962,Resort Hotel,1,352,2015,September,39,24,1,3,27,...,No Deposit,NaN,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
2003,Resort Hotel,1,354,2015,September,39,26,2,5,26,...,No Deposit,96.0,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
2164,Resort Hotel,1,361,2015,October,40,3,2,5,26,...,No Deposit,96.0,NaN,0,Group,0.0,0,0,Canceled,1/2/2015
2173,Resort Hotel,1,338,2015,October,41,4,2,0,55,...,No Deposit,NaN,NaN,0,Group,0.0,0,0,Canceled,1/2/2015


## Outlier - 3
I also found some bookings with unusually high numbers of adults.

After checking the customer type, I found that these were Group bookings and not data errors.

So, I decided to keep these records.

In [12]:
hbdc[
    (hbdc["adults"] == 0)
    &
    (hbdc["children"] == 0)
    &
    (hbdc["babies"] == 0)
]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
2224,Resort Hotel,0,1,2015,October,41,6,0,3,0,...,No Deposit,NaN,174.0,0,Transient-Party,0.00,0,0,Check-Out,10/6/2015
2409,Resort Hotel,0,0,2015,October,42,12,0,0,0,...,No Deposit,NaN,174.0,0,Transient,0.00,0,0,Check-Out,10/12/2015
3181,Resort Hotel,0,36,2015,November,47,20,1,2,0,...,No Deposit,38.0,NaN,0,Transient-Party,0.00,0,0,Check-Out,11/23/2015
3684,Resort Hotel,0,165,2015,December,53,30,1,4,0,...,No Deposit,308.0,NaN,122,Transient-Party,0.00,0,0,Check-Out,1/4/2016
3708,Resort Hotel,0,165,2015,December,53,30,2,4,0,...,No Deposit,308.0,NaN,122,Transient-Party,0.00,0,0,Check-Out,1/5/2016
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
115029,City Hotel,0,107,2017,June,26,27,0,3,0,...,No Deposit,7.0,NaN,0,Transient,100.80,0,0,Check-Out,6/30/2017
115091,City Hotel,0,1,2017,June,26,30,0,1,0,...,No Deposit,NaN,NaN,0,Transient,0.00,1,1,Check-Out,7/1/2017
116251,City Hotel,0,44,2017,July,28,15,1,1,0,...,No Deposit,425.0,NaN,0,Transient,73.80,0,0,Check-Out,7/17/2017
116534,City Hotel,0,2,2017,July,28,15,2,5,0,...,No Deposit,9.0,NaN,0,Transient-Party,22.86,0,1,Check-Out,7/22/2017


## Outlier - 4

Also, I found 180 bookings with 0 adults, children and babies.

As this is not possible, and could be there due to typing errors, I decided to drop these records.

In [13]:
hbdc[hbdc.duplicated()].head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
5,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.00,0,1,Check-Out,7/3/2015
22,Resort Hotel,0,72,2015,July,27,1,2,4,2,...,No Deposit,250.0,NaN,0,Transient,84.67,0,1,Check-Out,7/7/2015
43,Resort Hotel,0,70,2015,July,27,2,2,3,2,...,No Deposit,250.0,NaN,0,Transient,137.00,0,1,Check-Out,7/7/2015
138,Resort Hotel,1,5,2015,July,28,5,1,0,2,...,No Deposit,240.0,NaN,0,Transient,97.00,0,0,Canceled,7/1/2015
200,Resort Hotel,0,0,2015,July,28,7,0,1,1,...,No Deposit,240.0,NaN,0,Transient,109.80,0,3,Check-Out,7/8/2015


In [14]:
import pandas as pd
hbdc = pd.read_csv("C:\\Users\\varsh\\Desktop\\DA\\Hotel_booking_DC\\Data\\hotel_bookings_raw.csv.csv")

hbdc[hbdc.duplicated()].shape

(31994, 32)

In [15]:
round(hbdc.duplicated().sum() / len(hbdc)*100, 2)

np.float64(26.8)

In [16]:
hbdc[hbdc.duplicated()].sample(5)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
51935,City Hotel,0,316,2016,May,22,26,0,3,2,...,No Deposit,30.0,NaN,0,Transient-Party,110.00,0,0,Check-Out,5/29/2016
74107,City Hotel,1,328,2015,September,37,10,0,2,2,...,Non Refund,1.0,NaN,0,Contract,62.00,0,0,Canceled,1/1/2015
43157,City Hotel,0,76,2015,September,38,16,0,2,2,...,No Deposit,1.0,NaN,0,Transient-Party,62.00,0,0,Check-Out,9/18/2015
82531,City Hotel,0,72,2015,December,53,27,2,4,2,...,No Deposit,NaN,NaN,0,Transient,128.22,0,2,Check-Out,1/2/2016
45999,City Hotel,1,134,2015,December,49,4,2,2,2,...,Non Refund,3.0,NaN,91,Transient,65.00,0,0,Canceled,10/22/2015


In [17]:
dup_rows = hbdc[hbdc.duplicated(keep = False)]
dup_rows

dup_rows.shape

(40165, 32)

In [18]:
hbdc.drop_duplicates(inplace = True)

In [19]:
hbdc.shape

(87396, 32)

In [20]:
hbdc.duplicated().sum()

np.int64(0)

## Duplicate Records Investigation

I found 31,994 duplicate rows in the dataset.

Before removing them, I checked whether they were genuine records or exact duplicates.

Using duplicated(keep=False), I confirmed that the records were completely identical.

### Action Taken

Removed 31,994 duplicate rows.

### Result

Dataset size reduced from 119,390 rows to 87,396 rows.

In [21]:
hbdc.isnull().sum()

hotel                                 0
is_canceled                           0
lead_time                             0
arrival_date_year                     0
arrival_date_month                    0
arrival_date_week_number              0
arrival_date_day_of_month             0
stays_in_weekend_nights               0
stays_in_week_nights                  0
adults                                0
children                              4
babies                                0
meal                                  0
country                             452
market_segment                        0
distribution_channel                  0
is_repeated_guest                     0
previous_cancellations                0
previous_bookings_not_canceled        0
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
agent                             12193
company                           82137


In [22]:
round(hbdc.isnull().sum() / len(hbdc) * 100, 2)

hotel                              0.00
is_canceled                        0.00
lead_time                          0.00
arrival_date_year                  0.00
arrival_date_month                 0.00
arrival_date_week_number           0.00
arrival_date_day_of_month          0.00
stays_in_weekend_nights            0.00
stays_in_week_nights               0.00
adults                             0.00
children                           0.00
babies                             0.00
meal                               0.00
country                            0.52
market_segment                     0.00
distribution_channel               0.00
is_repeated_guest                  0.00
previous_cancellations             0.00
previous_bookings_not_canceled     0.00
reserved_room_type                 0.00
assigned_room_type                 0.00
booking_changes                    0.00
deposit_type                       0.00
agent                             13.95
company                           93.98


## Missing Values Analysis

After removing duplicates, I calculated the percentage of missing values.

### Findings

- Childre - 0.003%
- Country - 0.52%
- Agent - 13.95%
- Company - 93.98%

Each column was investigated separately before deciding on it.

In [23]:
hbdc["agent"].value_counts().head(20)

agent
9.0      28759
240.0    13028
14.0      3349
7.0       3300
250.0     2779
241.0     1644
28.0      1502
8.0       1383
1.0       1232
6.0       1117
40.0       986
314.0      844
242.0      722
83.0       614
85.0       524
243.0      477
171.0      402
27.0       395
3.0        363
22.0       345
Name: count, dtype: int64

### To know where the agent values are missing 

In [24]:
hbdc[hbdc["agent"].isnull()].head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.00,0,0,Check-Out,7/2/2015
6,Resort Hotel,0,0,2015,July,27,1,0,2,2,...,No Deposit,NaN,NaN,0,Transient,107.00,0,0,Check-Out,7/3/2015
18,Resort Hotel,0,0,2015,July,27,1,0,1,2,...,No Deposit,NaN,110.0,0,Transient,107.42,0,0,Check-Out,7/2/2015


In [25]:
hbdc["company"].value_counts().head(20)

company
40.0     851
223.0    503
45.0     238
153.0    206
154.0    133
219.0    131
174.0    121
281.0    119
233.0     95
51.0      80
405.0     77
94.0      76
47.0      62
331.0     60
169.0     53
135.0     52
110.0     48
91.0      46
62.0      44
67.0      44
Name: count, dtype: int64

In [26]:
hbdc["company"].nunique()

352

In [28]:
hbdc["company"].value_counts().head(25)

company
40.0     851
223.0    503
45.0     238
153.0    206
154.0    133
219.0    131
174.0    121
281.0    119
233.0     95
51.0      80
405.0     77
94.0      76
47.0      62
331.0     60
169.0     53
135.0     52
110.0     48
91.0      46
62.0      44
67.0      44
270.0     43
113.0     36
9.0       36
148.0     36
280.0     36
Name: count, dtype: int64

In [29]:
hbdc[hbdc["company"].notnull()].head(25)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
18,Resort Hotel,0,0,2015,July,27,1,0,1,2,...,No Deposit,NaN,110.0,0,Transient,107.42,0,0,Check-Out,7/2/2015
218,Resort Hotel,1,1,2015,July,28,8,0,1,2,...,No Deposit,NaN,110.0,0,Transient,104.72,0,1,Canceled,7/8/2015
220,Resort Hotel,0,1,2015,July,28,8,0,2,2,...,No Deposit,NaN,110.0,0,Transient,104.72,1,1,Check-Out,7/10/2015
295,Resort Hotel,0,84,2015,July,28,11,1,1,2,...,No Deposit,NaN,113.0,0,Transient,100.00,1,0,Check-Out,7/13/2015
387,Resort Hotel,0,1,2015,July,29,14,0,1,1,...,No Deposit,NaN,270.0,0,Transient,134.00,0,0,Check-Out,7/15/2015
450,Resort Hotel,1,12,2015,July,29,16,2,3,2,...,No Deposit,NaN,178.0,0,Transient,139.00,0,0,Canceled,7/7/2015
541,Resort Hotel,0,0,2015,July,30,19,1,0,2,...,No Deposit,NaN,240.0,0,Group,88.20,0,0,Check-Out,7/20/2015
600,Resort Hotel,0,5,2015,July,30,21,0,2,1,...,No Deposit,NaN,154.0,0,Transient,99.50,0,0,Check-Out,7/23/2015
628,Resort Hotel,1,19,2015,July,30,23,1,3,2,...,No Deposit,NaN,178.0,0,Transient,139.00,0,0,Canceled,7/7/2015
796,Resort Hotel,1,0,2015,July,31,30,0,1,1,...,No Deposit,NaN,270.0,0,Transient,114.00,0,0,Canceled,7/30/2015


In [30]:
hbdc["has_company"] = hbdc["company"].notnull().astype(int)

In [31]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,NaN,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,NaN,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,NaN,NaN,0,Transient,75.00,0,0,Check-Out,7/2/2015,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,304.0,NaN,0,Transient,75.00,0,0,Check-Out,7/2/2015,0
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,240.0,NaN,0,Transient,98.00,0,1,Check-Out,7/3/2015,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,394.0,NaN,0,Transient,96.14,0,0,Check-Out,9/6/2017,0
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,9.0,NaN,0,Transient,225.43,0,2,Check-Out,9/7/2017,0
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,9.0,NaN,0,Transient,157.71,0,4,Check-Out,9/7/2017,0
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,89.0,NaN,0,Transient,104.40,0,0,Check-Out,9/7/2017,0


In [32]:
hbdc.drop("company", axis = 1, inplace = True)

In [33]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,0,Transient,75.00,0,0,Check-Out,7/2/2015,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,0,Transient,75.00,0,0,Check-Out,7/2/2015,0
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,0,Transient,98.00,0,1,Check-Out,7/3/2015,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,No Deposit,394.0,0,Transient,96.14,0,0,Check-Out,9/6/2017,0
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,No Deposit,9.0,0,Transient,225.43,0,2,Check-Out,9/7/2017,0
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,No Deposit,9.0,0,Transient,157.71,0,4,Check-Out,9/7/2017,0
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,No Deposit,89.0,0,Transient,104.40,0,0,Check-Out,9/7/2017,0


## Company Column Investigation

The company column contained 93.98% missing values.

After checking the records, I found that company IDs represented bookings made through companies. Most missing values appeared to represent regular customer bookings rather than data errors.

### Action Taken

Created a new column:

- 1 = Booking through a company
- 0 = Regular booking

The original company column was then removed.

In [34]:
hbdc["agent"].value_counts().head(20)

agent
9.0      28759
240.0    13028
14.0      3349
7.0       3300
250.0     2779
241.0     1644
28.0      1502
8.0       1383
1.0       1232
6.0       1117
40.0       986
314.0      844
242.0      722
83.0       614
85.0       524
243.0      477
171.0      402
27.0       395
3.0        363
22.0       345
Name: count, dtype: int64

In [35]:
hbdc[hbdc["agent"].isnull()].head(20)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,0,Transient,75.00,0,0,Check-Out,7/2/2015,0
6,Resort Hotel,0,0,2015,July,27,1,0,2,2,...,No Deposit,NaN,0,Transient,107.00,0,0,Check-Out,7/3/2015,0
18,Resort Hotel,0,0,2015,July,27,1,0,1,2,...,No Deposit,NaN,0,Transient,107.42,0,0,Check-Out,7/2/2015,1
30,Resort Hotel,0,118,2015,July,27,1,4,10,1,...,No Deposit,NaN,0,Transient,62.00,0,2,Check-Out,7/15/2015,0
32,Resort Hotel,1,96,2015,July,27,1,2,8,2,...,No Deposit,NaN,0,Transient,108.30,0,2,Canceled,5/29/2015,0
42,Resort Hotel,0,16,2015,July,27,2,2,3,2,...,No Deposit,NaN,0,Transient,123.00,0,0,Check-Out,7/7/2015,0
55,Resort Hotel,0,1,2015,July,27,2,0,1,2,...,No Deposit,NaN,0,Transient,107.00,1,2,Check-Out,7/3/2015,0
56,Resort Hotel,0,0,2015,July,27,2,0,1,2,...,No Deposit,NaN,0,Transient,147.00,0,0,Check-Out,7/3/2015,0


In [36]:
hbdc["has_agent"] = hbdc["agent"].notnull().astype(int)

In [37]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,agent,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,NaN,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,NaN,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,304.0,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,240.0,0,Transient,98.00,0,1,Check-Out,7/3/2015,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,394.0,0,Transient,96.14,0,0,Check-Out,9/6/2017,0,1
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,9.0,0,Transient,225.43,0,2,Check-Out,9/7/2017,0,1
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,9.0,0,Transient,157.71,0,4,Check-Out,9/7/2017,0,1
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,89.0,0,Transient,104.40,0,0,Check-Out,9/7/2017,0,1


In [38]:
hbdc.drop("agent", axis = 1, inplace = True)

In [39]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,98.00,0,1,Check-Out,7/3/2015,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,No Deposit,0,Transient,96.14,0,0,Check-Out,9/6/2017,0,1
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,No Deposit,0,Transient,225.43,0,2,Check-Out,9/7/2017,0,1
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,157.71,0,4,Check-Out,9/7/2017,0,1
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,104.40,0,0,Check-Out,9/7/2017,0,1


In [40]:
hbdc.head(25)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,98.00,0,1,Check-Out,7/3/2015,0,1
6,Resort Hotel,0,0,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,107.00,0,0,Check-Out,7/3/2015,0,0
7,Resort Hotel,0,9,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,103.00,0,1,Check-Out,7/3/2015,0,1
8,Resort Hotel,1,85,2015,July,27,1,0,3,2,...,No Deposit,0,Transient,82.00,0,1,Canceled,5/6/2015,0,1
9,Resort Hotel,1,75,2015,July,27,1,0,3,2,...,No Deposit,0,Transient,105.50,0,0,Canceled,4/22/2015,0,1
10,Resort Hotel,1,23,2015,July,27,1,0,4,2,...,No Deposit,0,Transient,123.00,0,0,Canceled,6/23/2015,0,1


In [41]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,98.00,0,1,Check-Out,7/3/2015,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,No Deposit,0,Transient,96.14,0,0,Check-Out,9/6/2017,0,1
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,No Deposit,0,Transient,225.43,0,2,Check-Out,9/7/2017,0,1
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,157.71,0,4,Check-Out,9/7/2017,0,1
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,104.40,0,0,Check-Out,9/7/2017,0,1


## Agent Column Investigation

The agent column contained travel agent IDs.

After checking the missing values, I found that many bookings were made without an agent.

### Action Taken

Created a new column:

- 1 = Booking through an agent
- 0 = Booking without an agent

The original agent column was removed.

In [42]:
hbdc[hbdc["country"].isnull()].head(20)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
30,Resort Hotel,0,118,2015,July,27,1,4,10,1,...,No Deposit,0,Transient,62.0,0,2,Check-Out,7/15/2015,0,0
4127,Resort Hotel,1,0,2016,February,8,15,0,0,0,...,No Deposit,0,Transient,0.0,0,0,Canceled,2/15/2016,1,0
7092,Resort Hotel,1,8,2016,July,30,21,0,1,1,...,No Deposit,0,Transient,73.0,0,2,Canceled,7/20/2016,1,0
7860,Resort Hotel,1,39,2016,August,36,30,0,5,2,...,No Deposit,0,Transient,159.0,0,5,Canceled,7/22/2016,0,0
8779,Resort Hotel,1,0,2016,October,42,13,0,1,1,...,No Deposit,0,Transient,50.0,0,0,Canceled,10/13/2016,1,0
9376,Resort Hotel,1,0,2016,November,48,21,0,0,0,...,No Deposit,0,Group,0.0,0,0,Canceled,11/21/2016,1,0
9610,Resort Hotel,1,4,2016,December,52,20,0,1,1,...,No Deposit,0,Transient,30.0,0,2,Canceled,12/19/2016,1,0
11262,Resort Hotel,1,87,2017,April,18,30,2,3,2,...,No Deposit,0,Transient,48.0,0,0,Canceled,2/2/2017,0,1
13803,Resort Hotel,0,7,2016,January,3,13,0,2,1,...,No Deposit,0,Transient,27.0,0,0,Check-Out,1/15/2016,0,1
13804,Resort Hotel,0,7,2016,January,4,18,1,3,1,...,No Deposit,0,Transient,28.5,1,0,Check-Out,1/22/2016,1,0


In [43]:
hbdc["country"].value_counts().head(20)

country
PRT    27453
GBR    10433
FRA     8837
ESP     7252
DEU     5387
ITA     3066
IRL     3016
BEL     2081
BRA     1995
NLD     1911
USA     1875
CHE     1570
CN      1093
AUT      947
SWE      837
CHN      816
POL      765
RUS      561
NOR      515
ROU      458
Name: count, dtype: int64

In [44]:
hbdc["country"].nunique()

177

In [45]:
hbdc["country"] = hbdc["country"].fillna("Unknown")

In [46]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,98.00,0,1,Check-Out,7/3/2015,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,No Deposit,0,Transient,96.14,0,0,Check-Out,9/6/2017,0,1
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,No Deposit,0,Transient,225.43,0,2,Check-Out,9/7/2017,0,1
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,157.71,0,4,Check-Out,9/7/2017,0,1
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,104.40,0,0,Check-Out,9/7/2017,0,1


## Country Column Investigation

The country column had only 0.52% missing values.

I checked the missing records but could not find a reliable pattern to determine the correct countries.

### Action Taken

I replaced missing values with "Unknown" to retain all records.

In [47]:
hbdc[hbdc["children"].isnull()]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
40600,City Hotel,1,2,2015,August,32,3,1,0,2,...,No Deposit,0,Transient-Party,12.0,0,1,Canceled,8/1/2015,0,0
40667,City Hotel,1,1,2015,August,32,5,0,2,2,...,No Deposit,0,Transient-Party,12.0,0,1,Canceled,8/4/2015,0,1
40679,City Hotel,1,1,2015,August,32,5,0,2,3,...,No Deposit,0,Transient-Party,18.0,0,2,Canceled,8/4/2015,0,0
41160,City Hotel,1,8,2015,August,33,13,2,5,2,...,No Deposit,0,Transient-Party,76.5,0,1,Canceled,8/9/2015,0,1


In [48]:
hbdc["children"].describe()

count    87392.000000
mean         0.138640
std          0.455881
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max         10.000000
Name: children, dtype: float64

In [49]:
hbdc["children"] = hbdc["children"].fillna(hbdc["children"].median())

In [50]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,7/1/2015,0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,7/2/2015,0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,98.00,0,1,Check-Out,7/3/2015,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,No Deposit,0,Transient,96.14,0,0,Check-Out,9/6/2017,0,1
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,No Deposit,0,Transient,225.43,0,2,Check-Out,9/7/2017,0,1
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,157.71,0,4,Check-Out,9/7/2017,0,1
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,104.40,0,0,Check-Out,9/7/2017,0,1


## Children Column Investigation

The children column contained only 4 missing values.

As the number was very small and there was no enough information to determine the correct values, I used the median value.

### Action Taken

Filled missing values with the median value (0).

In [51]:
import pandas as pd
hbdc["reservation_status_date"] = pd.to_datetime(hbdc["reservation_status_date"])

## Datatype Conversion

The reservation_status_date column was stored as text.

### Action Taken

Converted the column from object datatype to datetime format for easier date analysis.

In [52]:
hbdc [ (hbdc["adults"] == 0) &
       (hbdc["children"] == 0) &
       (hbdc["babies"] == 0) ]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
2224,Resort Hotel,0,1,2015,October,41,6,0,3,0,...,No Deposit,0,Transient-Party,0.00,0,0,Check-Out,2015-10-06,1,0
2409,Resort Hotel,0,0,2015,October,42,12,0,0,0,...,No Deposit,0,Transient,0.00,0,0,Check-Out,2015-10-12,1,0
3181,Resort Hotel,0,36,2015,November,47,20,1,2,0,...,No Deposit,0,Transient-Party,0.00,0,0,Check-Out,2015-11-23,0,1
3684,Resort Hotel,0,165,2015,December,53,30,1,4,0,...,No Deposit,122,Transient-Party,0.00,0,0,Check-Out,2016-01-04,0,1
3708,Resort Hotel,0,165,2015,December,53,30,2,4,0,...,No Deposit,122,Transient-Party,0.00,0,0,Check-Out,2016-01-05,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
115029,City Hotel,0,107,2017,June,26,27,0,3,0,...,No Deposit,0,Transient,100.80,0,0,Check-Out,2017-06-30,0,1
115091,City Hotel,0,1,2017,June,26,30,0,1,0,...,No Deposit,0,Transient,0.00,1,1,Check-Out,2017-07-01,0,0
116251,City Hotel,0,44,2017,July,28,15,1,1,0,...,No Deposit,0,Transient,73.80,0,0,Check-Out,2017-07-17,0,1
116534,City Hotel,0,2,2017,July,28,15,2,5,0,...,No Deposit,0,Transient-Party,22.86,0,1,Check-Out,2017-07-22,0,1


In [53]:
hbdc = hbdc[
    ~(
        (hbdc["adults"] == 0) &
        (hbdc["children"] == 0) &
        (hbdc["babies"] == 0)
    )
]

In [54]:
hbdc

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,2015-07-01,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.00,0,0,Check-Out,2015-07-01,0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,2015-07-02,0,0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,0,Transient,75.00,0,0,Check-Out,2015-07-02,0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,0,Transient,98.00,0,1,Check-Out,2015-07-03,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119385,City Hotel,0,23,2017,August,35,30,2,5,2,...,No Deposit,0,Transient,96.14,0,0,Check-Out,2017-09-06,0,1
119386,City Hotel,0,102,2017,August,35,31,2,5,3,...,No Deposit,0,Transient,225.43,0,2,Check-Out,2017-09-07,0,1
119387,City Hotel,0,34,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,157.71,0,4,Check-Out,2017-09-07,0,1
119388,City Hotel,0,109,2017,August,35,31,2,5,2,...,No Deposit,0,Transient,104.40,0,0,Check-Out,2017-09-07,0,1


### Understanding "~" in Pandas

The "~" operator means "NOT" or "reverse the condition".

'df[condition]' keeps only the rows matching the condition, whereas 'df[~condition]' keeps all rows except those matching the condition.

"~condition = NOT these rows" else "Remove rows matching this condition."

## Outlier Treatment

I found 166 records where adults, children, and babies were all zero.

A hotel booking cannot realistically exist without any guests.

### Action Taken

Removed these invalid records.

In [55]:
hbdc[hbdc["adr"] < 0]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
14969,Resort Hotel,0,195,2017,March,10,5,4,6,2,...,No Deposit,0,Transient-Party,-6.38,0,0,Check-Out,2017-03-15,0,1


## Outlier Treatment

I found one booking with a negative ADR value (-6.38).

Since room prices cannot normally be negative, this record was treated as invalid.

### Action Taken

Removed the record.

In [56]:
hbdc = hbdc[hbdc["adr"] >= 0]

In [57]:
hbdc[hbdc["adr"] > 1000]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
48515,City Hotel,1,35,2016,March,13,25,0,1,2,...,Non Refund,0,Transient,5400.0,0,0,Canceled,2016-02-19,0,1


In [58]:
hbdc = hbdc[hbdc["adr"] <= 1000]

## Outlier Treatment

I found one booking with an ADR value of 5400.

This value was extremely different from the rest of the dataset and appeared unrealistic.

### Action Taken

Removed the record.

In [59]:
hbdc[ (hbdc["stays_in_weekend_nights"] == 0) &
      (hbdc["stays_in_week_nights"] == 0) ]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.0,0,0,Check-Out,2015-07-01,0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,0,Transient,0.0,0,0,Check-Out,2015-07-01,0,0
167,Resort Hotel,0,111,2015,July,28,6,0,0,2,...,No Deposit,0,Transient,0.0,0,2,Check-Out,2015-07-06,0,1
168,Resort Hotel,0,0,2015,July,28,6,0,0,1,...,No Deposit,0,Transient,0.0,0,0,Check-Out,2015-07-06,0,1
196,Resort Hotel,0,8,2015,July,28,7,0,0,2,...,No Deposit,0,Transient,0.0,0,1,Check-Out,2015-07-07,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
115483,City Hotel,0,15,2017,July,27,6,0,0,1,...,No Deposit,0,Transient-Party,0.0,0,0,Check-Out,2017-07-06,0,1
117701,City Hotel,0,0,2017,August,32,8,0,0,2,...,No Deposit,0,Transient,0.0,0,0,Check-Out,2017-08-08,0,1
118029,City Hotel,0,0,2017,August,33,14,0,0,2,...,No Deposit,0,Transient,0.0,0,0,Check-Out,2017-08-14,0,0
118631,City Hotel,0,78,2017,August,34,23,0,0,1,...,No Deposit,0,Transient-Party,0.0,0,0,Check-Out,2017-08-23,0,1


In [60]:
hbdc[
    (hbdc["stays_in_weekend_nights"] == 0) &
    (hbdc["stays_in_week_nights"] == 0)
]["reservation_status"].value_counts()

reservation_status
Check-Out    568
No-Show       13
Canceled      10
Name: count, dtype: int64

### Outlier Treatment

I checked bookings where both weekend nights and weekday nights were zero.

After investigation, many of these bookings were marked as Check-Out, Canceled, or No-Show.

Since these records can occur in real hotel operations, they were retained.

In [62]:
len(hbdc)

87228

In [63]:
hbdc.shape

(87228, 32)

In [64]:
hbdc[
    (hbdc["adults"] == 0) &
    (hbdc["children"] == 0) &
    (hbdc["babies"] == 0)
].shape

(0, 32)

In [65]:
hbdc.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests   

In [69]:
hbdc.shape

(87228, 32)

In [70]:
hbdc.duplicated().sum()

np.int64(17)

In [71]:
hbdc[
    (hbdc["adults"] == 0) &
    (hbdc["children"] == 0) &
    (hbdc["babies"] == 0)
].shape

(0, 32)

In [72]:
hbdc[hbdc["adr"] < 0]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent


In [73]:
hbdc[hbdc["adr"] > 1000]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,has_company,has_agent


In [74]:
hbdc = hbdc.drop_duplicates()

In [75]:
hbdc.duplicated().sum()

np.int64(0)

In [76]:
hbdc.shape

(87211, 32)

In [77]:
hbdc.to_csv(
    "C:\\Users\\varsh\\Desktop\\DA\\Hotel_booking_DC\\Data\\hotel_bookings_cleaned.csv",
    index=False
)

# Data Cleaning Summary

### Actions Performed

- Removed 31,994 duplicate records
- Filled missing values in the children column
- Replaced missing country values with "Unknown"
- Created the has_company feature
- Created the has_agent feature
- Converted reservation_status_date to datetime
- Removed invalid bookings with zero guests
- Removed one negative ADR record
- Removed one extremely high ADR record

### Final Dataset Shape

(87,211 rows, 32 columns)

The dataset is now cleaned and ready for Exploratory Data Analysis (EDA), SQL analysis, and dashboard creation.